### ISSUE TITLE
Reach For the Sphere segmentation fault 

### ISSUE THREAD

**Issue by nissmar** [sgsellan/gpytoolbox#100]
Hey! 

I'm trying to run _reach for the sphere_, but the code crashes on some models:
```
V,F = gpy.read_mesh("53159.obj")
j = 32
sdf = lambda x: gpy.signed_distance(x, V, F)[0]
gx, gy, gz = np.meshgrid(np.linspace(-1.0, 1.0, j+1), np.linspace(-1.0, 1.0, j+1), np.linspace(-1.0, 1.0, j+1))
U = np.vstack((gx.flatten(), gy.flatten(), gz.flatten())).T
V0, F0 = gpy.icosphere(2)
Vr,Fr = gpy.reach_for_the_spheres(U, sdf, V0, F0, min_h=.01)
```

Output:

```
 Segmentation fault (core dumped)
```

Is there some requirements on the input? The one I used (53159 of the Thingi10k dataset) is watertight and normalized to -0.8/0.8, but its genus is not zero.

**Comment by sgsellan** [comment]
Hi! Thanks for using our code!

Indeed our method will never change the topology of the shape, so if it starts from a sphere (that's what V0, F0 are), it will only be able to recover genus 0 shapes. We'll make this clearer in the documentation.

**Comment by nissmar** [comment]
Thank you for your quick reply! 
I understand that the genus of the output will be zero but from the article I had the impression that reach_for_the_spheres could still deform the input mesh to best fit the target shape (for instance if the target is a cup and the input a sphere, the latter could somehow be deformed to fit the cup with a "filled" handle
). Is that not the case?

**Comment by sgsellan** [comment]
I see! No, that's not the case, if the input is genus 1, the flow will attempt to change topology and reach a singularity, leading to a crash (see Fig. 19 in our paper). Nonetheless, we should catch these singularities better so that there's no segfault. I'll try to replicate your error on my side to fix it.

**Comment by sgsellan** [comment]
Hi Nissmar! 

I spent a little bit longer debugging this and tracked down the issue. Indeed, because our method is not intended to work if the input mesh and sdf genus don't agree, singularities were being caused; nonetheless, we should catch these singularities inside our code and return your last converged result so that the method fails gracefully. It should never segfault. There was an issue with how we stitched together the active and inactive mesh after each flow iteration that caused this segfault in extreme cases. It is fixed in #101 now.

If you install gpytoolbox from that PR branch and run the updated algorithm (I don't have access to your `53159.obj`, so I use the [stl file linked here](https://thingiverse-production-new.s3.amazonaws.com/assets/c5/90/46/fb/49/mushroom.stl) by doing
```python
V,F = gpy.read_mesh("test/unit_tests_data/53159.stl")
# is mesh normalized? print corners
# print(np.min(V, axis=0))
# print(np.max(V, axis=0))
V = gpy.normalize_points(V)
 j = 32
sdf = lambda x: gpy.signed_distance(x, V, F)[0]
gx, gy, gz = np.meshgrid(np.linspace(-1.0, 1.0, j+1), np.linspace(-1.0, 1.0, j+1), np.linspace(-1.0, 1.0, j+1))
U = np.vstack((gx.flatten(), gy.flatten(), gz.flatten())).T
V0, F0 = gpy.icosphere(2)
Vr,Fr = gpy.reach_for_the_spheres(U, sdf, V0, F0, min_h = .01, verbose = False)
```
the algorithm encounters a singularity, but now catches it inside the python code and returns the following ugly-yet-intended mesh:
<img width="574" alt="Screenshot 2023-11-06 at 8 55 23 AM" src="https://github.com/sgsellan/gpytoolbox/assets/30415788/fdd84ecf-3b3d-4afb-9765-995e4acd4737">

You can even see the singularity forming like a little black hole in the top left :) intuitively, that is the method "trying" to get the topology change without succeeding.

Let me know if you have any more questions! Best of luck in your pursuit of the spheres.

- Silvia

**Comment by nissmar** [comment]
Hi Silvia! Thank you very much for your reply and fixing that segmentation fault, I'm sure it will be useful for other people as well! Thank you also for the insights on your great method :)
